include/ringbuffer_utils.hpp

```diff
@@ -152,6 +154,160 @@ class PerfRingBufferReader {
   uint64_t _head;
 };
 
+struct MPSCRingBufferHeader {
+  uint64_t size;
+  static constexpr uint64_t k_discard_bit = 1UL << 62;
+  static constexpr uint64_t k_busy_bit = 1UL << 63;
+
+  static bool is_busy(uint64_t size) { return size & k_busy_bit; }
+  static bool is_discarded(uint64_t size) { return size & k_discard_bit; }
+};
+
+class MPSCRingBufferWriter {
+public:
+  static inline constexpr std::chrono::milliseconds k_lock_timeout{100};
+
+  explicit MPSCRingBufferWriter(RingBuffer &rb) : _rb(rb) {
+    assert(rb.type == RingBufferType::kMPSCRingBuffer);
+    update_tail();
+  }
+
+  MPSCRingBufferWriter(const MPSCRingBufferWriter &) = delete;
+  MPSCRingBufferWriter &operator=(const MPSCRingBufferWriter &) = delete;
+
+  void update_tail() {
+    _tail = __atomic_load_n(_rb.reader_pos, __ATOMIC_ACQUIRE);
+  }
+
+  Buffer reserve(size_t n, bool *timeout = nullptr) {
+    size_t n2 = align_up(n + sizeof(MPSCRingBufferHeader), 8);
+    if (n2 == 0) {
+      return {};
+    }
+
+    // \fixme{nsavoire} Not sure if spinlock is the best option here
+    if (!lock(reinterpret_cast<std::atomic<bool> *>(_rb.spinlock),
+              k_lock_timeout)) {
+      // timeout on lock
+      if (timeout) {
+        *timeout = true;
+      }
+      return {};
+    }
+
+    // No need for atomic operation, since we hold the lock
+    uint64_t writer_pos = *_rb.writer_pos;
+
+    uint64_t new_writer_pos = writer_pos + n2;
+    // Check that there is enough free space
+    if (_rb.mask < new_writer_pos - _tail) {
+      unlock(reinterpret_cast<std::atomic<bool> *>(_rb.spinlock));
+      return {};
+    }
+
+    uint64_t head_linear = writer_pos & _rb.mask;
+    MPSCRingBufferHeader *hdr =
+        reinterpret_cast<MPSCRingBufferHeader *>(_rb.data + head_linear);
+
+    // Mark the sample as busy
+    hdr->size = n | MPSCRingBufferHeader::k_busy_bit;
+
+    // Atomic operation required to synchronize with reader load_acquire
+    __atomic_store_n(_rb.writer_pos, new_writer_pos, __ATOMIC_RELEASE);
+    unlock(reinterpret_cast<std::atomic<bool> *>(_rb.spinlock));
+
+    return {reinterpret_cast<std::byte *>(hdr) + sizeof(MPSCRingBufferHeader),
+            n};
+  }
+
+  // Return true if notification to consumer is necesssary
+  // Notification is necessary only if consumer has caught up with producer
+  // (meaning tail afer commit is at or after head before commit)
+  bool commit(Buffer buf, bool discard = false) {
+    MPSCRingBufferHeader *hdr = reinterpret_cast<MPSCRingBufferHeader *>(
+        buf.data() - sizeof(MPSCRingBufferHeader));
+
+    // Clear busy bit
+    uint64_t new_size = hdr->size ^ MPSCRingBufferHeader::k_busy_bit;
+    if (discard) {
+      new_size |= MPSCRingBufferHeader::k_discard_bit;
+    }
+
+    __atomic_store_n(&hdr->size, new_size, __ATOMIC_RELAXED);
```

Comment:
Ooops, `relaxed` is not good here, since we need that all previous writes to be visible to the reader once we change `hdr->size`. Need to be changed to `release`